### PR TITLE
Derive debug and opaque types

### PR DESCRIPTION
**Pull request by fitzgen**
See each commit for details.

Follow up to #842.

r? @emilio or @Manishearth

### CODE REVIEW

**Comment by fitzgen** [comment]
I think I may need to update some libclang version specific tests once I get word back from Travis CI...

File: src/ir/analysis/derive_debug.rs [hunk 1]
```diff
@@ -138,6 +137,17 @@ impl<'ctx, 'gen> MonotoneFramework for CannotDeriveDebug<'ctx, 'gen> {
             Some(ty) => ty
         };
 
+        if ty.is_opaque(self.ctx, item) {
+            let layout_can_derive = ty.layout(self.ctx).map_or(true, |l| {
+                l.opaque().can_trivially_derive_debug(self.ctx, ())
+            });
+            if layout_can_derive {
```
Comment (emilio) [inline src/ir/analysis/derive_debug.rs]: nit: `return if`?

File: tests/headers/opaque-template-inst-member.hpp [hunk 2]
```diff
@@ -3,10 +3,18 @@
 template<typename T>
 class OpaqueTemplate {
     T mData;
-    bool mCannotDebug[40];
+    bool mCannotDebug[400];
 };
 
+// This should not end up deriving Debug because its `mBlah` field cannot derive
+// Debug because the instantiation's definition cannot derive Debug.
 class ContainsOpaqueTemplate {
```
Comment (emilio) [inline tests/headers/opaque-template-inst-member.hpp]: Part of the point of the test is that we derived debug for the instantiation if we could, could you add that back, maybe as another different test-case?

File: tests/expectations/tests/opaque-template-inst-member-2.rs [hunk 3]
```diff
@@ -0,0 +1,67 @@
+/* automatically generated by rust-bindgen */
+
+
+#![allow(dead_code, non_snake_case, non_camel_case_types, non_upper_case_globals)]
```
Comment (emilio) [inline tests/expectations/tests/opaque-template-inst-member-2.rs]: Oh, sure, disregard my previous comment :-)

**Comment by fitzgen** [comment]
@bors-servo r=emilio

Thanks for review :)

**Pull request merged**